File: transfers/legit/ledger/builder.py
from dataclasses import dataclass

from transfers.family.engine import GraphConfig, TransferConfig
from transfers.factory import TransactionFactory

from .limits import build_balance_book
from .passes import add_credit, add_family, add_income, add_routines
from .screenbook import ScreenBook
from .streams import TxnStreams

from transfers.legit.blueprints import (
    Blueprint,
    TransfersPayload,
    build_legit_plan,
)


@dataclass(slots=True)
class LegitTransferBuilder:
    request: Blueprint
    fam_graph_cfg: GraphConfig
    fam_transfer_cfg: TransferConfig

    def payload(self) -> TransfersPayload:
        if not self.request.network.accounts.ids:
            return TransfersPayload(
                candidate_txns=[],
                hub_accounts=[],
                biller_accounts=[],
                employers=[],
                initial_book=None,
            )

        plan = build_legit_plan(
            self.request.timeline,
            self.request.network,
            self.request.macro,
            self.request.overrides,
        )

        initial_book = build_balance_book(
            self.request.timeline,
            self.request.network,
            self.request.specs,
            self.request.cc_state,
            plan,
        )

        txf = TransactionFactory(
            rng=self.request.timeline.rng,
            infra=self.request.overrides.infra,
        )

        streams = TxnStreams()
        screen = ScreenBook(initial_book)

        add_income(self.request, plan, txf, streams)
        add_routines(self.request, plan, txf, streams, screen)
        add_family(
            self.request,
            plan,
            txf,
            streams,
            self.fam_graph_cfg,
            self.fam_transfer_cfg,
        )
        add_credit(self.request, plan, txf, streams)

        return TransfersPayload(
            candidate_txns=streams.candidates,
            hub_accounts=plan.counterparties.hub_accounts,
            biller_accounts=plan.counterparties.biller_accounts,
            employers=plan.counterparties.employers,
            initial_book=initial_book,
            replay_sorted_txns=streams.replay_ready,
        )

    def build(self) -> TransfersPayload:
        return self.payload()

File: src/transfers/legit/ledger/builder.cpp
#include "phantomledger/transfers/legit/ledger/builder.hpp"

#include "phantomledger/transactions/factory.hpp"
#include "phantomledger/transfers/legit/blueprints/plans.hpp"
#include "phantomledger/transfers/legit/ledger/limits.hpp"
#include "phantomledger/transfers/legit/ledger/passes.hpp"
#include "phantomledger/transfers/legit/ledger/screenbook.hpp"
#include "phantomledger/transfers/legit/ledger/streams.hpp"

#include <stdexcept>

namespace PhantomLedger::transfers::legit::ledger {

namespace {

[[nodiscard]] const entity::account::Registry *
accountsFrom(const LegitTransferRequest &request) noexcept {
  return request.plan.census.accounts;
}

[[nodiscard]] const relationships::family::Households &
familyHouseholdsFrom(const LegitTransferRequest &request) noexcept {
  return request.familyHouseholds != nullptr
             ? *request.familyHouseholds
             : relationships::family::kDefaultHouseholds;
}

[[nodiscard]] const relationships::family::Dependents &
familyDependentsFrom(const LegitTransferRequest &request) noexcept {
  return request.familyDependents != nullptr
             ? *request.familyDependents
             : relationships::family::kDefaultDependents;
}

[[nodiscard]] const relationships::family::RetireeSupport &
retireeSupportFrom(const LegitTransferRequest &request) noexcept {
  return request.retireeSupport != nullptr
             ? *request.retireeSupport
             : relationships::family::kDefaultRetireeSupport;
}

} // namespace

TransfersPayload LegitTransferBuilder::build() const {
  if (request == nullptr) {
    throw std::invalid_argument(
        "LegitTransferBuilder.build() requires a non-null request");
  }

  const auto *accounts = accountsFrom(*request);
  if (accounts == nullptr || accounts->records.empty()) {
    return TransfersPayload{};
  }

  auto plan = blueprints::buildLegitPlan(request->plan);

  auto initialBook = buildBalanceBook(request->balanceBook, plan);

  TxnStreams streams;
  ScreenBook screen{initialBook.get()};

  if (request->plan.rng == nullptr) {
    throw std::invalid_argument(
        "LegitTransferBuilder.build() requires a non-null rng");
  }
  const transactions::Factory txf(*request->plan.rng, request->router);

  passes::GovernmentCounterparties govCps{};

  passes::addIncome(request->income, plan, txf, streams, govCps);

  if (request->plan.census.ownership != nullptr &&
      request->plan.census.accounts != nullptr) {
    passes::addRoutines(request->routines, plan,
                        *request->plan.census.ownership,
                        *request->plan.census.accounts, txf, streams, screen);
  }

  if (request->familyTransfers != nullptr) {
    passes::addFamily(request->family, plan, txf, streams,
                      familyHouseholdsFrom(*request),
                      familyDependentsFrom(*request),
                      retireeSupportFrom(*request), *request->familyTransfers);
  }

  passes::addCredit(request->credit, plan, txf, streams);

  // Payload packing.
  TransfersPayload payload;
  payload.candidateTxns = streams.takeCandidates();
  payload.hubAccounts = std::move(plan.counterparties.hubAccounts);
  payload.billerAccounts = std::move(plan.counterparties.billerAccounts);
  payload.employers = std::move(plan.counterparties.employers);
  payload.initialBook = std::move(initialBook);
  payload.replaySortedTxns = streams.takeReplayReady();

  return payload;
}

} // namespace PhantomLedger::transfers::legit::ledger
